### ISSUE TITLE
Should we use the `if_rust_version` crate?

### ISSUE THREAD

**Issue by Kixunil** [rust-bitcoin/rust-bitcoin#3211]
I've just noticed [a crate](https://docs.rs/if_rust_version/1.0.0/if_rust_version/macro.if_rust_version.html) that does pretty much the exact same thing I implemented recently. It might be worth using it because if any downstream consumer uses it it will save compile times even better. I have no knowledge of the authors and their reputation but the crate is trivial to review and likely in `crev` already. It's also very stable and has crazy low MSRV, so the risks are really low.

We could also have it as a feature on `internals` and suggest people to turn it on if they want to reduce compile times (but that'll still require compiling an empty `build.rs`)

**Comment by tcharding** [comment]
Should be easy enough to review, only 100 lines of code and 200 in the build file.

**Comment by Kixunil** [comment]
FTR I've already read it in its entirety, just not in maximum suspicion brain mode. :) I'm pretty sure it doesn't access unrelated files or something obvious like that.

**Comment by apoelstra** [comment]
Yeah, I'd be fine with it (with the caveat that I haven't read the code and have yet to). I like that he's almost apologetic for not having a lower MSRV than 1.12. Though I guess when the last version was released 5 years ago that was maybe a recent version.

Funny that his syntax is the same as yours.

I think that if it elimates our build.rs in internals that's enough of a win to be worthwhile. I would really like if we could just get rid of build.rs everywhere (except rust-secp ... and even then I wonder if we can avoid it somehow in the case that we link to the system library, when we support that).

**Comment by Kixunil** [comment]
> I think that if it elimates our build.rs in internals that's enough of a win to be worthwhile. I would really like if we could just get rid of build.rs everywhere

Note that it still just kicks the can down the road - then `if_rust_version` will have the build script. But hopefully more widely reviewed one and shared among more crates so its costs gets reduced. We could also make it optional with the caveat that turning it off will simply remove all features that require a newer version regardless of the actual version being used.

> and even then I wonder if we can avoid it somehow in the case that we link to the system library, when we support that

Yes it is possible IFF a) we don't use `bindgen` or similar-non-compilation/linking-related stuff (thankfully, we don't) and b) we don't do any funny hacks and use the standard stuff provided by the language - we fail on this one for having a different library name in `Cargo.toml`. I was thinking about reporting these as candidates for getting a more of a first-class support from the language but didn't find time to do it.

**Comment by apoelstra** [comment]
> Note that it still just kicks the can down the road - then if_rust_version will have the build script. 

True, but it's narrowly scoped (and guaranteed to stay narrowly scoped because the crate only does one thing).

> We could also make it optional with the caveat that turning it off will simply remove all features that require a newer version regardless of the actual version being used.
> 

oooo, I like this.

**Comment by Kixunil** [comment]
> True, but it's narrowly scoped (and guaranteed to stay narrowly scoped because the crate only does one thing).

Exactly, that's what I like about it. A serious project could probably hard pin it by having a check that `Cargo.lock` contains specific hash to avoid backdoor surprises without much issues.

**Comment by tcharding** [comment]
> We could also make it optional with the caveat that turning it off will simply remove all features that require a newer version regardless of the actual version being used.

One potential problem would be if we needed both `if` and `else` clauses, it may be hard to write something that works with all compilers and features on and off. We don't tend to use both clauses though.

**Comment by Kixunil** [comment]
What does it mean to need both of them? You can write both and `else` only makes sense when you're writing two conflicting things (e.g. `const` condition). So if the feature is off we just always emit `no`.

**Comment by apoelstra** [comment]
Right -- conceptually "feature off" is the same as "the compiler is the MSRV".

**Comment by tcharding** [comment]
I agree that binary MSRV  or not works. But I was just flagging that finer grained control may not be possible if the macro is always feature gated.

As an example:

```
#[cfg(feature = "versioning-feature"]
{
  rust_version_macro! {
    if >= 1.70 { /* 1st code here */ } else { /* 2nd code here */ }
 }
}
```

How would we get the 2nd code to run if all calls to the macro are feature gated? I.e compiler is greater than MSRV but less than some other version. Oh, answering my own question, if the macro supports logical and (`&&`) we can.

**Comment by apoelstra** [comment]
The *calls* to the macro aren't feature-gated. The content of the macro is.

So without the feature-gate, we have a dummy macro implementation that always uses the `else` clause. *With* the feature gate we do the clever trick we do now where we selectively use the `else` clause.

**Comment by Kixunil** [comment]
We just reexport our own dumb version of macro that just always emits the no branch.
